cowbird/permissions_synchronizer.py

```diff
@@ -493,8 +531,11 @@ def _find_permissions_to_sync(self, src_res_key, src_matched_groups, input_permi
 
         return permission_data
 
-    def sync(self, perm_operation, permission, src_resource_tree):
-        # type: (Callable[[List[Dict]], None], Permission, List[Dict]) -> None
+    def sync(self,
+             perm_operation: Callable[[List[PermissionConfigItemType]], None],
+             permission: Permission,
+             src_resource_tree: ResourceTree,
```

Comment:
Not sure of the `ResourceTree` type, used here and in `create_permission`/`delete_permission`.

`ResourceTree` is defined as `List[Dict[str, ResourceSegment]]` and 
`ResourceSegment = TypedDict("ResourceSegment", {"resource_name": str, "resource_type": str})`

The variable here seems to be a list containing items such as this dict :
```
{'resource_name': 'catalog', 
'resource_display_name': 'catalog', 
'resource_type': 'service', 
'resource_id': 22, 
'parent_id': None, 
'root_service_id': None, 
'permission_names': ['browse-match', 'browse-allow-match', ...], 
'permissions': [{'name': 'browse', 'access': 'allow', 'scope': 'match', 'type': 'allowed'}, ...]}
```

We might need a different type distinct of `ResourceSegment` to distinguish between the dict found here and the data used for example in `_create_resdata()` which only uses the `resource_name` and `resource_type` keys. Or maybe there is already a distinct type in Magpie?

Also, I think it might be `List[ResourceSegment]` instead of `List[Dict[str, ResourceSegment]]`.

Reply:
In Magpie, the return type is just a generic `JSON`. The `cast` in the other places `ResourceTree` is employed are applied because those functions return `List[JSON]`. Therefore, even a more detailed type definition will require a `cast` to enforce a more explicit mapping.

I don't mind having a more explicit type, but currently, the only place `ResourceTree` items are accessed is in this `sync` method, and it only accesses `resource_name` and `resource_type`. 

For now, I'll leave it as is because the types are "good enough", but I'll add a note about adjusting with some eventual Magpie type if it gets implemented. I don't want to replicate too much the response structure of Magpie in Cowbird because we cannot be warned properly of invalid types if Magpie types evolves and Cowbird redefines (possibly incorrectly) its own types.